bad k=v and no check_raw is an error

bcoca opened the pull request:
fixes #84756

##### ISSUE TYPE

- Bugfix Pull Request

webknjaz commented on lib/ansible/utils/vars.py:
```diff
@@ -198,7 +198,10 @@ def load_extra_vars(loader):
+                try:
+                    data = parse_kv(extra_vars_opt)
+                except AnsibleParserError as e:
+                    raise AnsibleOptionsError("Invalid extra vars option passed", orig_exc=e)
```
```suggestion
                    raise AnsibleOptionsError(
                        "Invalid extra vars option passed",
                        orig_exc=e,
                    ) from e
```

bcoca replied:
this would add the message x2 once we merge `from e` functions to devel, one or the other, not both.

webknjaz commented on lib/ansible/parsing/splitter.py:
Use `raise .. from` in both places here.

bcoca replied:
since this is intended for backport and that is not currently working, keeping the orig_exc. I do intend to update them all to `from e` once DT is merged.

mattclay commented:
Please update `test/units/parsing/test_splitter.py` to maintain the 100% code coverage it provides on `lib/ansible/parsing/splitter.py`.

mattclay commented on test/units/parsing/test_splitter.py:
```diff
@@ -146,6 +146,10 @@
 PARSE_KV = tuple((test[0], test[2]) for test in SPLIT_DATA)
+REMAINDER_ERROR = (
+    (u'creates=something thisisextra'),
+    (u'something thisisextra'),
```
The `u` prefix is unnecessary.

mattclay replied:
Also, since these values are only used once, just include them inline for the test which uses them:

```python
@pytest.mark.parametrize("args", (
    'creates=something thisisextra',
    'something thisisextra',
))
```

bcoca replied:
I knew `u` was not needed, I just try to keep the style of existing code

mattclay commented on test/units/parsing/test_splitter.py:
```diff
@@ -173,3 +177,9 @@ def test_split_args_error(args):
+
+
+@pytest.mark.parametrize("args", REMAINDER_ERROR)
+def test_parse_kv_remainder_error(args):
```
Include type hints on new code, including unit tests, when reasonable:

```suggestion
def test_parse_kv_remainder_error(args: str) -> None:
```

mattclay commented on test/units/parsing/test_splitter.py:
```diff
@@ -173,3 +177,9 @@ def test_split_args_error(args):
+
+@pytest.mark.parametrize("args", REMAINDER_ERROR)
+def test_parse_kv_remainder_error(args):
+    with pytest.raises(ValueError):
```
Verify that the `ValueError` encountered is the correct one, and not due to an unexpected failure:

```suggestion
    with pytest.raises(ValueError, match='Invalid k=v detected'):
```

mattclay commented on test/units/parsing/test_splitter.py:
```diff
@@ -173,3 +173,9 @@ def test_split_args_error(args):
         parse_kv(args)
+
+
+@pytest.mark.parametrize("args", ('creates=something thisisextra', 'something thisisextra'))
```
Keep the parameters to one-per-line so it's easier to read and add new ones:

```suggestion
@pytest.mark.parametrize("args", (
    'creates=something thisisextra', 
    'something thisisextra',
))
```

mattclay reviewed (commented):
The error handling for `Invalid k=v detected` needs some improvement.

For example `ansible -m ping localhost -e 'ansible_facts_parallel: False'` yields:

```
ERROR! Unexpected Exception, this is probably a bug: Invalid k=v detected: ansible_facts_parallel:
```

bcoca commented:
@mattclay, to parse as YAML you need `{ }`, so the complaint is correct

mattclay commented on test/integration/targets/parsing/runme.sh:
```diff
@@ -4,3 +4,9 @@ set -eux
+
+# test that we don't inject _raw_params with invalid extra vars
+[ "$(ansible-playbook no_extras.yml -i ../../inventory -e 'bad var as string' 2>&1 |grep -c 'Invalid k=v detected')" -gt "0" ]
+[ "$(ansible-playbook no_extras.yml -i ../../inventory -e 'valid=var inkvalidstring' 2>&1 |grep -c 'Invalid k=v detected')" -gt "0" ]
```
Update the regex to include the error message generated when catching the `ValueError`. This will ensure the exception handler is working as intended. Also, since the playbook here never runs, just use an ad-hoc task:

```suggestion
[ "$(ansible -m debug testhost -i ../../inventory -e 'bad var as string' 2>&1 |grep -c 'Invalid extra vars option passed. Invalid k=v detected')" -gt "0" ]
[ "$(ansible -m debug testhost -i ../../inventory -e 'valid=var inkvalidstring' 2>&1 |grep -c 'Invalid extra vars option passed. Invalid k=v detected')" -gt "0" ]
```

mattclay commented on test/integration/targets/parsing/no_extras.yml:
```diff
@@ -0,0 +1,7 @@
+  - name: ensure bad extra vars parameter does not create bad var
+    assert:
+      that:
+        - _raw_params is not defined
```
Include an assertion for the `-e` value which is parsed:

```suggestion
        - _raw_params is not defined
        - valid == "as string"
```

mattclay commented on test/integration/targets/parsing/runme.sh:
```diff
@@ -4,3 +4,9 @@ set -eux
+[ "$(ansible-playbook no_extras.yml -i ../../inventory -e 'bad var as string' 2>&1 |grep -c 'Invalid k=v detected')" -gt "0" ]
+[ "$(ansible-playbook no_extras.yml -i ../../inventory -e 'valid=var inkvalidstring' 2>&1 |grep -c 'Invalid k=v detected')" -gt "0" ]
+# but we can still create normally
+ansible-playbook no_extras.yml -i ../../inventory -e 'valid="as string"' "$@"
```
With the above change to ad-hoc, the `no_extras.yml` can be renamed to more accurately reflect it's purpose. Perhaps call it `valid_extras.yml` instead?

mattclay commented on lib/ansible/parsing/mod_args.py:
```diff
@@ -182,7 +182,7 @@ def _normalize_parameters(self, thing, action=None, additional_args=None):
                 tmp_args = args.pop('args')
                 if isinstance(tmp_args, string_types):
-                    tmp_args = parse_kv(tmp_args)
+                    tmp_args = parse_kv(tmp_args, check_raw=bool(action in FREEFORM_ACTIONS))
```
What's this change for? I noticed it was in the original commit, along with changelog entries that appear to be related, but those changelog entries were removed in a later commit.

webknjaz replied:
Also, a `__contains__()` check already returns a `bool`. No need to convert bool to bool.
```suggestion
                    tmp_args = parse_kv(tmp_args, check_raw=action in FREEFORM_ACTIONS)
```

mattclay commented on lib/ansible/parsing/splitter.py:
```diff
@@ -82,8 +82,10 @@ def parse_kv(args, check_raw=False):
+            elif remainder:
                 raw_params.append(orig_x)
+            else:
+                raise ValueError(f"Invalid k=v detected: {orig_x}")
```
Thinking ahead to when DT is merged, we'll want this to be an `AnsibleError` derived type. That will allow us to set `obj=args`, which will provide error message context when this function is used in Jinja plugins. Without DT, setting `obj` will have no effect, so there should be no issue including it.

Speaking of Jinja plugins, shouldn't the `sequence` lookup plugin's usage of `parse_kv` set `remainder=False`, since `_raw_params` is not one of the fields it supports?

mattclay commented on lib/ansible/parsing/splitter.py:
```diff
@@ -44,7 +44,7 @@ def decode_match(match):
 
 
-def parse_kv(args, check_raw=False):
+def parse_kv(args, check_raw=False, remainder=True):
```
I think renaming `remainder` to something more descriptive, perhaps `allow_raw_params`, would make it more clear what it's used for. What do you think?

webknjaz replied:
It's a good idea to rename the arg. May I also ask you to update the docstring to mention it as well?

webknjaz commented on changelogs/fragments/extras_avoid_freform.yml:
```diff
@@ -0,0 +1,3 @@
+bugfixes:
+  - invalid extra vars will not create a `_raw_params` variable with the invalid string as value anymore.
+  - parse_kv function will only return a `_raw_params` when the caller asks it to via new `remainder` option.
```
```suggestion
  - invalid extra vars will not create a ``_raw_params`` variable with the invalid string as value anymore.
  - ``parse_kv()`` function will only return a ``_raw_params`` when the caller asks it to via new ``remainder`` argument.
```

webknjaz commented on test/integration/targets/gathering_facts/runme.sh:
```diff
@@ -12,7 +12,7 @@ ANSIBLE_GATHERING=smart ansible-playbook test_run_once.yml -i inventory -v "$@"
 
 # ensure fact merging is working properly
-ansible-playbook verify_merge_facts.yml -v "$@" -e 'ansible_facts_parallel: False'
+ansible-playbook verify_merge_facts.yml -v "$@" -e '{ansible_facts_parallel: False}'
```
Could you explain this change?

bcoca replied:
it was not creating the correct variable, my 'fix' pointed out this error, which was going unoticed due to the bug

webknjaz commented on test/units/parsing/test_splitter.py:
```diff
@@ -173,3 +173,12 @@ def test_split_args_error(args):
+@pytest.mark.parametrize("args", (
+    'creates=something thisisextra',
+    'something thisisextra',
+))
```
Could you add `ids=` here to make it obvious what the semantics of the failing cases is?

bcoca replied:
?

mattclay replied:
In most cases test IDs are unnecessary. They're mostly useful when the parametrized values don't repr well -- which is not the case here. Since there's only a single string parameter, IDs don't really add anything here.

webknjaz commented on test/units/parsing/test_splitter.py:
```diff
@@ -173,3 +173,12 @@ def test_split_args_error(args):
+    'something thisisextra',
+))
+def test_parse_kv_remainder_error(args: str) -> None:
+    with pytest.raises(ValueError, match='Invalid k=v detected'):
```
Let's make it obvious that it's not a simple substring match.
```suggestion
    with pytest.raises(ValueError, match=r'^Invalid k=v detected: .*$'):
```

mattclay commented on test/units/parsing/test_splitter.py:
```diff
@@ -173,3 +173,12 @@ def test_split_args_error(args):
+    'something thisisextra',
+))
+def test_parse_kv_remainder_error(args: str) -> None:
+    with pytest.raises(ValueError, match=r'^Invalid k=v detected: .*$'):
```
It's probably not a bad idea to verify the string starts with the expected message, but it's pointless to check the end of the string, so this can be simplified:

```suggestion
    with pytest.raises(ValueError, match=r'^Invalid k=v detected: '):
```

mattclay commented on lib/ansible/parsing/mod_args.py:
```diff
@@ -182,7 +182,7 @@ def _normalize_parameters(self, thing, action=None, additional_args=None):
                 tmp_args = args.pop('args')
                 if isinstance(tmp_args, string_types):
-                    tmp_args = parse_kv(tmp_args)
+                    tmp_args = parse_kv(tmp_args, check_raw=action in FREEFORM_ACTIONS)
```
What's this change for? I noticed it was in the original commit, along with changelog entries that appear to be related, but those changelog entries were removed in a later commit.

bcoca replied:
that should now affect the new parameter, as _raw_params for a module that does not support them should be an error

mattclay commented on lib/ansible/parsing/splitter.py:
```diff
@@ -44,12 +44,12 @@ def decode_match(match):
-    to a new parameter called '_raw_params'. If check_raw is not enabled,
-    they will simply be ignored.
+    are found and the check_raw is enabled they can be passed back as is.
+    If populate_raw_paramjs option is set to True, they will be added
```
```suggestion
    If populate_raw_params option is set to True, they will be added
```

mattclay commented on lib/ansible/parsing/splitter.py:
```diff
@@ -44,12 +44,12 @@ def decode_match(match):
-    they will simply be ignored.
+    are found and the check_raw is enabled they can be passed back as is.
+    If populate_raw_paramjs option is set to True, they will be added
+    to a new parameter called '_raw_params' otherwise they return an error.
```
```suggestion
    to a new parameter called '_raw_params' otherwise an AnsibleParserError is raised.
```

mattclay commented on lib/ansible/plugins/lookup/sequence.py:
```diff
@@ -139,7 +139,7 @@ class LookupModule(LookupBase):
     """
 
-    def parse_kv_args(self, args):
+    def parse_kv_args(self, args, populate_raw_params=False):
```
Why is this change here? The new argument is unused.

This is the `parse_kv` usage I was referring to in an earlier review comment:

https://github.com/ansible/ansible/blob/5347d4d4fcdd4ec53aa272ca208d837f395baaef/lib/ansible/plugins/lookup/sequence.py#L230

mattclay commented on test/units/parsing/test_splitter.py:
```diff
@@ -173,3 +173,12 @@ def test_split_args_error(args):
+    'something thisisextra',
+))
+def test_parse_kv_remainder_error(args: str) -> None:
+    with pytest.raises(ValueError, match=r'^Invalid k=v detected: '):
```
This needs to be `AnsibleParserError` now:

```suggestion
    with pytest.raises(AnsibleParserError, match=r'^Invalid k=v detected: '):
```

mattclay commented on lib/ansible/plugins/lookup/sequence.py:
```diff
@@ -227,7 +227,7 @@ def run(self, terms, variables, **kwargs):
                 try:
                     if not self.parse_simple_args(term):
-                        self.parse_kv_args(parse_kv(term))
+                        self.parse_kv_args(parse_kv(term, populate_raw_params=False))
```
Can you add an integration test to verify the error handling for this?

bcoca replied:
was already on it!

bcoca replied:
it seems we would error out later on either way, this just fails faster and slightly changes the message.

bcoca commented:
rethinking this as 2 PRs:
-  one for backport, detect _raw_params in extra and throw exception
- one going forward, new parse_kv where `check_raw` really means `check_raw` and optional exception parameter on 'fail_on_partial_parse', deprecate existing in favor of new

bcoca committed:
extra vars should turn unparsed into _raw_params

This makes it an error if the full string is parsed as 'raw'
but a warning if it is only a partial remainder

bcoca commented on lib/ansible/utils/vars.py:
```diff
@@ -199,6 +201,11 @@ def load_extra_vars(loader):
+                if '_raw_params' in data:
+                    display.warning(f"Ignoring unparsable data: {data['_raw_params']}")
+                    del data['_raw_params']
+                    if not data:  # if only bad args, reset to none to fail below
```
i'm 50/50 on this being a warning vs hard error
